docs(website): add pagination doc page (#1407)
twilio-labs/paste: packages/paste-website/src/pages/components/pagination/index.mdx

import {Blockquote} from '../../../components/Blockquote';
import {defaultExample, numberExample, positioningExample} from '../../../component-examples/PaginationExamples';

export const pageQuery = graphql`
  {
    allPasteComponent(filter: {name: {eq: "@twilio-paste/pagination"}}) {
      edges {
        node {
          name
          description
          status
          version
        }
      }
    }
    mdx(fields: {slug: {eq: "/components/pagination/"}}) {
      fileAbsolutePath
      frontmatter {
        slug
        title
      }
      headings {
        depth
        value
      }
    }
  }
`;

<ComponentHeader
  name={props.pageContext.frontmatter.title}
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/pagination"
  storybookUrl="/?path=/story/components-pagination--arrows"
  abstractUrl="https://www.figma.com/file/E6KUvMhioUmAgN0nwmReTM/Paste-Components?node-id=1469%3A0"
  data={props.data.allPasteComponent.edges}
/>

---

<contentwrapper>

<PageAside data={props.data.mdx} />

<content>

# Guidelines

## About Pagination

> Use Pagination to split up content when a user’s goal is "to find a specific item [in a list] and click through to that destination page" (source: [Nielsen Norman Group](https://www.nngroup.com/articles/item-list-view-all/)). Pagination is used to communicate only where a user is in paged content. It isn’t used to show the status of each page- for that, a Progress Stepper ([coming soon](/roadmap)) would be used.
> 
> Pagination is often paired with [Tables](/components/table), one of the most commonly used components in Twilio and one of the primary ways customers view content in our products. Think of Pagination as an important navigation element that helps our customers understand exactly where they are and help them decide where to go next.

Reviewer: ```suggestion
Use Pagination to split up content when a user’s goal is "to find a specific item [in a list] and click through to that destination page" (source: [Nielsen Norman Group](https://www.nngroup.com/articles/item-list-view-all/)). Pagination is used to communicate only where a user is in paged content. It isn’t used to show the status of each page—use a Progress Stepper ([coming soon](/roadmap)) instead.

Pagination is often paired with [Tables](/components/table), one of the most commonly used components in Twilio and one of the primary ways customers view content in our products. Think of Pagination as an important navigation element that helps our customers understand exactly where they are and help them decide where to go next.

The component is made up of 3 main parts:
- `PaginationArrow`s: Left and right arrow controls, with an option to add text labels
- `PaginationLabel`: Indicates the current page in view (e.g., "Page 2", "1–10 of 50 results")
- `PaginationNumbers`: Shows a list of page numbers and the total page count
```

    <PaginationItems>
      <PaginationArrow label="Go to previous month" variant="back" visibleLabel="Previous" />
      <PaginationLabel>February 2021</PaginationLabel>
      <PaginationArrow label="Go to next month" variant="forward" visibleLabel="Next" />
    </PaginationItems>
  </Pagination>
  <Text as="span" display="block" fontStyle="italic" marginTop="space70">Will the "Next" button take you to January or March?</Text>
</>`}
</LivePreview>

To give customers further clarity on where they are in a data set, you can add custom text to:

- The current page label
- The left and right direction controls

<LivePreview scope={{Pagination, PaginationItems, PaginationArrow, PaginationLabel}} language="jsx">
  {`<Pagination label="current page label pagination navigation">
  <PaginationItems>
    <PaginationArrow label="Go to newer templates" variant="back" visibleLabel="Newer templates" />
    <PaginationLabel>51–100 of 500+ templates</PaginationLabel>
    <PaginationArrow label="Go to older templates" variant="forward" visibleLabel="Older templates" />
  </PaginationItems>
</Pagination>`}
</LivePreview>

### Pagination with total page count

> Use the Pagination variant with total page count in cases when knowing the exact size of a dataset is crucial to a customer’s understanding of the content (e.g., knowing how many phone numbers aren’t compliant) and when you’re able to fetch the data in a performant way. This variant also allows customers to skip ahead pages in a dataset.

Reviewer: ```suggestion
Use the Pagination variant with total page count in cases when knowing the exact size of a dataset is crucial to a customer’s understanding of the content (e.g., knowing how many phone numbers aren’t compliant) and when you’re able to fetch the data in a performant way.

This variant also allows customers to skip ahead pages in a dataset, though this hasn't been identified as a common use case.
```

<Callout>
  <CalloutTitle>Considering the tradeoffs for showing the total count of a dataset</CalloutTitle>
  <CalloutText>
    By default,{' '}

>     <Anchor href="https://www.twilio.com/blog/replacing-absolute-paging-and-related-properties">

Reviewer: ```suggestion
    <Anchor href="https://www.twilio.com/blog/replacing-absolute-paging-and-related-properties" showExternal>
```

      Twilio APIs might not support showing current page number or the total count of a dataset
    </Anchor>
    .
  </CalloutText>
  <CalloutText>
    However, anecdotal evidence from product designers suggests that knowing the size of a dataset is important to
    Twilio customers in certain cases. Showing the total count of a dataset makes sense especially when the customer has
    more active control over total count, such as with total SIMs, but not for datasets that rapidly increase, like
    event logs. If the exact total count is unknown, consider indicating size in another way that gives customers an
    approximate idea of the size of a dataset, like showing "1–10 of 100+ results".
  </CalloutText>
</Callout>

<LivePreview
  scope={{
    Pagination,
    PaginationItems,
    PaginationArrow,
    PaginationNumbers,
    PaginationNumber,
    Table,
    THead,
    TBody,
    Th,
    Tr,
    Td,
    Box,
  }}
  noInline
>
  {numberExample}

> </LivePreview>

Reviewer: In this example, can you show what the interaction would look like when there are ellipses? The ellipsis component's shown in the Usage Guide and Anatomy sections but mentioned nowhere else.

I'm not really sure how to use it (and if we'd need to add more guidelines around it).

Author: Added a section to the composition notes section with guidance and examples.

## Composition Notes

> Make sure the current page label and direction controls in the Pagination component give customers enough clarity on where they are and where they’ll be going. Show the current page number whenever possible.

Reviewer: ```suggestion
Make sure the current page label and arrow controls in the Pagination component give customers enough clarity on where they are and where they’ll be going. Show the current page number whenever possible.
```

### Positioning Pagination

The Pagination component should be centered underneath the content it controls separated by a margin of `$space-70`.

<LivePreview
  scope={{
    Pagination,
    PaginationItems,
    PaginationArrow,
    PaginationNumbers,
    PaginationNumber,
    Table,
    THead,
    TBody,
    Th,
    Tr,
    Td,
    Box,
    Text,
  }}
  noInline
>
  {positioningExample}
</LivePreview>

### Paging through chronological data

> Add custom text to direction controls (e.g., "Older"/"Newer", "Previous month"/"next month") for datasets sorted by time or date. This makes it clear what direction the pages are moving.
> 
> If a customer can change the sort order of the dataset, make sure you swap the labelling of the direction controls, as well.

Reviewer: ```suggestion
Add custom text to the arrow controls (e.g., "Older"/"Newer", "Previous month"/"next month") for datasets sorted by time or date. This makes it clear what direction the pages are moving.

If a customer can change the sort order of the dataset, make sure you swap the labelling of the arrow controls, as well.
```

If a single default can’t satisfy customers’ needs, you can give customers a way to change the number of items in view. Try to keep these options limited to reduce cognitive overhead, especially if the choice isn’t critical to the customer’s main goal.

<Blockquote
  name="Jakob Nielsen"
  source="Nielsen Norman Group"
  sourceUrl="https://www.nngroup.com/articles/item-list-view-all/"
>
  Many sites let users choose how many items they’ll see on each page. This is often overkill...It’s usually better to
  offer a single default number — such as 10 or 20 — [or] give users the choice between two numbers, say 10 and 50,
  where the second number is substantially bigger than the default. If the choice is between two relatively similar
  numbers (such as 10 and 20), users might as well click the Next Page button rather than suffer the cognitive overhead
  of trying to decide their display preference.
</Blockquote>

Consider [contributing a Table Actions pattern](/patterns/contribute) to help standardize this across Twilio!

## When to use Pagination

> 
> Use Pagination to split up content in a list where users are trying to find specific items in that list, like a Table or page of Cards.

Reviewer: ```suggestion
```

#### PaginationArrow Props

| Prop          | Type      | Description                                             | Default  |
| ------------- | --------- | ------------------------------------------------------- | -------- |
| as?           | `string`  | `a` / `button`                                          | `button` |
| disabled?     | `boolean` | Sets `aria-hidden` to be true, and `visibility: hidden` | null     |
| href?         | `string`  | href used when as `a`                                   | null     |
| label         | `string`  | `aria-label` text                                       | null     |
| variant       | `string`  | `back` / `forward`                                      | null     |
| visibleLabel? | `string`  | Visible text of the button or anchor                    | null     |

#### PaginationNumbers Props

| Prop       | Type        | Description                               | Default |
| ---------- | ----------- | ----------------------------------------- | ------- |
| children   | `ReactNode` |                                           | null    |
| pageLabel? | `string`    | Page label text used in `PaginationLabel` | null    |

> 

Reviewer: Is this section missing the PaginationNumber component with the `as?` prop? Is `as=a` set on PaginationNumber or PaginationNumbers?

> #### PaginationArrow Props

Reviewer: Is this supposed to be PaginationNumber?

Author: Yep. Good catch!

import {Blockquote} from '../../../components/Blockquote';
import {defaultExample, numberExample, positioningExample} from '../../../component-examples/PaginationExamples';

export const pageQuery = graphql`
  {
    allPasteComponent(filter: {name: {eq: "@twilio-paste/pagination"}}) {
      edges {
        node {
          name
          description
          status
          version
        }
      }
    }
    mdx(fields: {slug: {eq: "/components/pagination/"}}) {
      fileAbsolutePath
      frontmatter {
        slug
        title
      }
      headings {
        depth
        value
      }
    }
  }
`;

<ComponentHeader
  name={props.pageContext.frontmatter.title}
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/pagination"
  storybookUrl="/?path=/story/components-pagination--arrows"
  abstractUrl="https://www.figma.com/file/E6KUvMhioUmAgN0nwmReTM/Paste-Components?node-id=1469%3A0"
  data={props.data.allPasteComponent.edges}
/>

---

<contentwrapper>

<PageAside data={props.data.mdx} />

<content>

# Guidelines

## About Pagination

> Use Pagination to split up content when a user’s goal is "to find a specific item [in a list] and click through to that destination page" (source: [Nielsen Norman Group](https://www.nngroup.com/articles/item-list-view-all/)). It’s used to communicate only where a user is in paged content. It isn’t used to show the status of each page, which is what a Progress Stepper ([coming soon](/roadmap)) would be used for.

Reviewer 2: ```suggestion
Use Pagination to split up content when a user’s goal is "to find a specific item [in a list] and click through to that destination page" (source: [Nielsen Norman Group](https://www.nngroup.com/articles/item-list-view-all/)). Pagination is used to communicate only where a user is in paged content. It isn’t used to show the status of each page- for that, a Progress Stepper ([coming soon](/roadmap)) would be used.
```

- Use the Pagination component primarily when it controls a large part of the page in view, like a full-width Table. Since Pagination typically appears underneath the UI element it’s controlling, it needs to be obvious that it’s controlling the element above it, rather than anything else below it. For example, if you need to let users page through a small table in a card on a dashboard page, consider letting them expand the table and show Pagination on the expanded view instead.

## Examples

### Default pagination

Use the default Pagination component for paged datasets that frequently increase in size, like event logs, or where the total number of pages is unknown.

Whenever possible, show the _current_ page number, so that users know where they are in a dataset and get confirmation that they’ve moved to a new page.

<LivePreview scope={{Pagination, PaginationItems, PaginationArrow, PaginationLabel}} noInline disabled>
  {defaultExample}
</LivePreview>

#### Default pagination with Anchor

Use the Anchor variant for Pagination if the URL changes for each page.

> 

Reviewer 2: About Button vs. Anchor: I think it would be useful if PaginationArrow and the props that determine whether it is a Button or Anchor are stated more clearly before the LivePreview here. I feel like it's especially important to do it since PaginationArrow defaults to Button, and I wasn't sure if the heading's mention of "Default" was about the Anchor/Button or Pagination itself.

Also, since there is a `variant` prop that is not about anchor vs. button, I think it'd be better to say something like "Use the Anchor **version** of Pagination Arrow by setting the `as="a"` prop and including an `href` prop."

Author: @gloriliale good suggestion. I added a little more to that section. Let me know if you think that works.

  {`<Pagination label="current page label pagination navigation">
  <PaginationItems>
    <PaginationArrow label="Go to newer templates" variant="back" visibleLabel="Newer templates" />
    <PaginationLabel>51–100 of 500+ templates</PaginationLabel>
    <PaginationArrow label="Go to older templates" variant="forward" visibleLabel="Older templates" />
  </PaginationItems>
</Pagination>`}
</LivePreview>

### Pagination with total page count

Use the Pagination variant with total page count in cases when knowing the exact size of a dataset is crucial to a customer’s understanding of the content (e.g., knowing how many phone numbers aren’t compliant) and when you’re able to fetch the data in a performant way. This variant also allows customers to skip ahead pages in a dataset.

<Callout>
  <CalloutTitle>Considering the tradeoffs for showing the total count of a dataset</CalloutTitle>
  <CalloutText>
    By default,{' '}
    <Anchor href="https://www.twilio.com/blog/replacing-absolute-paging-and-related-properties">
      Twilio APIs might not support showing current page number or the total count of a dataset
    </Anchor>
    .
  </CalloutText>
  <CalloutText>

>     However, anecdotal evidence from product designers have suggested that knowing the size of a dataset is important to

Reviewer 2: ```suggestion
    However, anecdotal evidence from product designers suggests that knowing the size of a dataset is important to
```

    Twilio customers in certain cases. Showing the total count of a dataset makes sense especially when the customer has
    more active control over total count, such as with total SIMs, but not for datasets that rapidly increase, like
    event logs. If the exact total count is unknown, consider indicating size in another way that gives customers an
    approximate idea of the size of a dataset, like showing "1–10 of 100+ results".

>   </CalloutText>

Reviewer 2: This is a great callout, and I think it's in the right place for people to read it and get the info they need to make a good choice for their interface. Woohoo!

Most Twilio datasets show 10 or 50 rows per page by default. However, a generic rule doesn't always work for complex, context-dependent data. We recommend using research to reach a sensible default and solution that works best for the customer.

If a single default can’t satisfy customers’ needs, you can give customers a way to change the number of items in view. Try to keep these options limited to reduce cognitive overhead, especially if the choice isn’t critical to the customer’s main goal.

<Blockquote
  name="Jakob Nielsen"
  source="Nielsen Norman Group"
  sourceUrl="https://www.nngroup.com/articles/item-list-view-all/"
>
  Many sites let users choose how many items they’ll see on each page. This is often overkill...It’s usually better to
  offer a single default number — such as 10 or 20 — [or] give users the choice between two numbers, say 10 and 50,
  where the second number is substantially bigger than the default. If the choice is between two relatively similar
  numbers (such as 10 and 20), users might as well click the Next Page button rather than suffer the cognitive overhead
  of trying to decide their display preference.
</Blockquote>

> Consider [contributing a Table Actions pattern](/patterns/contribute) to help standardize this across Twilio!

Reviewer 2: Heck yeah!

  storybookUrl="/?path=/story/components-pagination--arrows"
  abstractUrl="https://www.figma.com/file/E6KUvMhioUmAgN0nwmReTM/Paste-Components?node-id=1469%3A0"
  data={props.data.allPasteComponent.edges}
/>

---

<contentwrapper>

<PageAside data={props.data.mdx} />

<content>

# Guidelines

## About Pagination

Use Pagination to split up content when a user’s goal is "to find a specific item [in a list] and click through to that destination page" (source: [Nielsen Norman Group](https://www.nngroup.com/articles/item-list-view-all/)). Pagination is used to communicate only where a user is in paged content. It isn’t used to show the status of each page—use a Progress Stepper ([coming soon](/roadmap)) instead.

Pagination is often paired with [Tables](/components/table), one of the most commonly used components in Twilio and one of the primary ways customers view content in our products. Think of Pagination as an important navigation element that helps our customers understand exactly where they are and help them decide where to go next.

The component is made up of 6 main parts:

- `Pagination`: The outer pagination container with the accessible `role=navigation`

> - `PaginationItems`: A wrapper that displays the pagination contens inline.

Reviewer: ```suggestion
- `PaginationItems`: A wrapper that displays the pagination contents inline.
```

> - `PaginationItems`: A wrapper that displays the pagination contens inline.
> - `PaginationArrow`s: Left and right arrow controls, with an option to add text labels

Reviewer: ```suggestion
- `PaginationItems`: A wrapper that displays the pagination contents inline.
- `PaginationArrow`: Left and right arrow controls, with an option to add text labels
```

    event logs. If the exact total count is unknown, consider indicating size in another way that gives customers an
    approximate idea of the size of a dataset, like showing "1–10 of 100+ results".
  </CalloutText>
</Callout>

<LivePreview
  scope={{
    Pagination,
    PaginationItems,
    PaginationArrow,
    PaginationNumbers,
    PaginationNumber,
    Table,
    THead,
    TBody,
    Th,
    Tr,
    Td,
    Box,
  }}
  noInline
>
  {numberExample}
</LivePreview>

## Composition Notes

Make sure the current page label and arrow controls in the Pagination component give customers enough clarity on where they are and where they’ll be going. Show the current page number whenever possible.

### Positioning Pagination

The Pagination component should be centered underneath the content it controls separated by a margin of `$space-70`.

<LivePreview
  scope={{
    Pagination,
    PaginationItems,
    PaginationArrow,
    PaginationNumbers,
    PaginationNumber,
    Table,
    THead,
    TBody,
    Th,
    Tr,
    Td,
    Box,
    Text,
  }}
  noInline
>
  {positioningExample}
</LivePreview>

### Using PaginationEllipsis

> When the number of pages exceeds seven, `PaginationEllipsis` can be used to indicate there are remaining pages.

Reviewer: ```suggestion
When the total number of pages exceeds seven, `PaginationEllipsis` is used to indicate there are remaining pages. Use `PaginationEllipsis` when the current page is more than 3 pages away from the first or last page.
```
